### ISSUE TITLE
Mix-up between directive and module injectors causes module injectables to be instantiated with directive-provided dependencies

### ISSUE THREAD

**Issue by matthewjh** [angular/angular#41582]
# 🐞 bug report

### Affected Package
@angular/core

### Is this a regression?

I don't know.

### Description

It appears possible for Angular to mix up directive and module injectors and incorrectly use the directive injector to provide dependencies for a module-level injectable. 

One way this happens is in the following flow:

Directive inject -> directive useFactory provider -> useFactory delegates to module-level injectable -> module-level injectable calls `Injector#get`

Assuming the subject of the `Injector#get` call is a tree-shakeable-provided service that has NOT already been created, then Angular enters the generated `ɵfac` function for the injectable -- correctly -- but the `ɵɵinject` calls embedded in the factory incorrectly delegate to the directive injector rather than to the module injector.

Below is the complete flow, indicating injector boundaries:

Directive Injector {_Directive inject -> directive useFactory provider -> useFactory delegates to module-level injectable ->_} Module Injector { _module-level injectable calls `Injector#get`  -> enter generated `ɵfac` for injected service -> **generated factory calls `ɵɵinject` for service dependencies**_ }

What appears to me to be happening is the last part in bold is delegated wrongly to the directive injector, irrespective of the earlier crossing of the boundary into module injector world.

This produces a broken state, as a "permanent" module-level injectable is created with something that belongs to a directive ("temporary").

## 🔬 Minimal Reproduction

https://github.com/matthewjh/angular-injector-repro

run

`ng serve`

go to app

open console

observe 

`tokenValue in A = from component`

as `A` is a providedIn: "root" injectable, the log output should be

`tokenValue in A = from module`

`app.component.ts`

```ts
import { Component, Inject, Injectable, InjectionToken, Injector } from '@angular/core';

const TOKEN = new InjectionToken<string>("string token", { providedIn: "root", factory: () => "from module"});
const A_TOKEN = new InjectionToken<A>("A");

@Injectable({providedIn: "root"})
class A {
  constructor (@Inject(TOKEN) tokenValue: string) {
    console.log(`tokenValue in A = ${tokenValue}`);
  }
}


@Injectable({providedIn: "root"})
class AGetter {
  constructor (private readonly injector: Injector) {}

  getA () {
    return this.injector.get(A);
  }
}


@Component({
  selector: 'app-root',
  templateUrl: './app.component.html',
  styleUrls: ['./app.component.css'],
  providers: [
    {provide: A_TOKEN, useFactory: (aGetter: AGetter) => aGetter.getA(), deps:[AGetter] },
    {provide: TOKEN, useValue: "from component"}
  ]
})
export class AppComponent {
  title = 'my-app';
  constructor (@Inject(A_TOKEN) a: A) {
  }
}
```

Issues that don't have enough info and can't be reproduced will be closed.

You can read more about issue submission guidelines here: https://github.com/angular/angular/blob/master/CONTRIBUTING.md#submit-issue
-->

## 🔥 Exception or Error
<pre><code>

</code></pre>

## 🌍  Your Environment

**Angular Version:**
    "@angular/common": "~11.2.9",
    "@angular/compiler": "~11.2.9",
    "@angular/core": "~11.2.9",
    "@angular/forms": "~11.2.9",
    "@angular/platform-browser": "~11.2.9",
    "@angular/platform-browser-dynamic": "~11.2.9",
    "@angular/router": "~11.2.9",

</code></pre>

**Anything else relevant?**

**Comment by matthewjh** [comment]
Further observations:

`setInjectImplementation` is called with `ɵɵdirectiveInject` in `getNodeInjectable`, before entering my `useFactory` provider on the directive:

`{provide: A_TOKEN, useFactory: (aGetter: AGetter) => aGetter.getA(), deps:[AGetter] },`

This is fine, but the problem is that `setInjectImplementation` is not called again when `A` calls `this.injector.get(A);`

This means that for the `ɵɵinject` call in 

`A.ɵfac = function A_Factory(t) { return new (t || A)(_angular_core__WEBPACK_IMPORTED_MODULE_0__["ɵɵinject"](TOKEN)); };`

`getInjectImplementation() ` returns  `ɵɵdirectiveInject`, so the directive injector is used to resolve the lookup.

**Comment by drummerboof** [comment]
Any word on this at all? I have run into this as well.

**Comment by AndrewKushnir** [comment]
FYI, there is a high chance that this problem is resolved by PR #42886. The changes were released to NPM as a part of the v12.1.3 version today. Please try to update to the v12.1.3 version and check the behavior once again. Thank you.
